src/engine/QueryPlanner.cpp

```diff
@@ -646,32 +643,82 @@ vector<QueryPlanner::SubtreePlan> QueryPlanner::getOrderByRow(
+        // There is an edge between *it->second and the node with id "id".
+        tg._adjLists[addedNode._id].push_back(otherNode._id);
+        tg._adjLists[otherNode._id].push_back(addedNode._id);
+      }
+    }
+  }
+}
+
 // _____________________________________________________________________________
 QueryPlanner::TripleGraph QueryPlanner::createTripleGraph(
     const p::BasicGraphPattern* pattern) const {
   TripleGraph tg;
-  if (pattern->_triples.size() > 64) {
-    AD_THROW("At most 64 triples allowed at the moment.");
-  }
+  size_t numNodesInTripleGraph = 0;
+  ad_utility::HashMap<Variable, string> optTermForCvar;
+  ad_utility::HashMap<Variable, vector<string>> potentialTermsForCvar;
+  vector<const SparqlTriple*> entityTriples;
+  // Add one or more nodes for each triple.
   for (auto& t : pattern->_triples) {
-    // Add a node for the triple.
-    tg._nodeStorage.emplace_back(TripleGraph::Node(tg._nodeStorage.size(), t));
-    auto& addedNode = tg._nodeStorage.back();
-    tg._nodeMap[addedNode._id] = &tg._nodeStorage.back();
-    tg._adjLists.emplace_back(vector<size_t>());
-    assert(tg._adjLists.size() == tg._nodeStorage.size());
-    assert(tg._adjLists.size() == addedNode._id + 1);
-    // Now add an edge between the added node and every node sharing a var.
-    for (auto& addedNodevar : addedNode._variables) {
-      for (size_t i = 0; i < addedNode._id; ++i) {
-        auto& otherNode = *tg._nodeMap[i];
-        if (otherNode._variables.count(addedNodevar) > 0) {
-          // There is an edge between *it->second and the node with id "id".
-          tg._adjLists[addedNode._id].push_back(otherNode._id);
-          tg._adjLists[otherNode._id].push_back(addedNode._id);
+    if (t._p._iri == CONTAINS_WORD_PREDICATE) {
+      string s = t._o.toString();
+      std::vector<std::string> terms = std::vector<std::string>(
+          absl::StrSplit(s.substr(1, s.size() - 2), ' '));
+      // Add one node for each word
+      for (size_t i = 0; i < terms.size(); i++) {
+        potentialTermsForCvar[t._s.getVariable()].push_back(terms[i]);
+        addNodeToTripleGraph(TripleGraph::Node(tg._nodeStorage.size(),
+                                               t._s.getVariable(), terms[i], t),
+                             tg);
+        numNodesInTripleGraph++;
+      }
+    } else if (t._p._iri == CONTAINS_ENTITY_PREDICATE) {
+      entityTriples.push_back(&t);
+    } else {
+      addNodeToTripleGraph(TripleGraph::Node(tg._nodeStorage.size(), t), tg);
+      numNodesInTripleGraph++;
+    }
+  }
+  for (const SparqlTriple* t : entityTriples) {
+    Variable currentVar = t->_s.getVariable();
+    if (!optTermForCvar.contains(currentVar)) {
+      if (potentialTermsForCvar.contains(currentVar)) {
+        optTermForCvar[currentVar] =
+            potentialTermsForCvar[currentVar]
+                                 [_qec->getIndex().getIndexOfBestSuitedElTerm(
+                                     potentialTermsForCvar[currentVar])];
```

Comment:
This becomes much simpler if you first go over the `potential...` to create all the `opt`...
Then for each triple you only simply check in the `opt..` and else throw. The code then gets much more linear instead of deeply nested.